doc/Port-Mac-Security/PMS.md

```diff
@@ -0,0 +1,309 @@
+    * [Configuration and Management Requirements](#configuration-and-management-requirements)
+    * [Scalability Requirements](#scalability-requirements)
+    * [Warm Boot Requirements](#warm-boot-requirements)
+  * [Functionality](#-functionality)
+    * [Functional Description](#functional-description)
+  * [Design](#design)
+    * [Overview](#overview)
+      * [PMS Configurations](#pms-configurations)
+      * [PMS Show Command](#pms-show-command)
+    * [DB Changes](#db-changes)
+      * [CONFIG DB](#config-db)
+      * [APP DB](#app-db)
+      * [COUNTERS DB](#counters-db)
+    * [SAI](#sai)
+      * [New Enum for violation action](#new-enum-for-violation-action)
+      * [New attributes in sai_port_attr_t](#new-attributes-in-sai_port_attr_t)
+    * [YANG model](#yang-model)
+      * [Configurations](#configurations)
+      * [Show](#show)
+  * [Flow Diagrams](#flow-diagrams)
+  * [Serviceability and Debug](#serviceability-and-debug)
+  * [Warm Boot Support](#warm-boot-support)
+  * [Scalability](#scalability)
+  * [Industry standard CLI](#industry-standard-cli)
+  * [Unit Test](#unit-test)
+
+# List of Tables
+
+[Table 1: Abbreviations](#table-1-abbreviations)
+
+# Revision
+
+| Rev  |    Date    |      Author      | Change Description |
+| :--: | :--------: | :--------------: | ------------------ |
+| 0.1  | 09/22/2021 | Anil Pandey      | Initial version    |
+
+# About this Manual
+
+This document provides general information about the Port Mac Security feature enhancement in SONiC.
+
+# Scope
+
+This document describes the high level design of Port Mac Security feature. 
+
+# Definition/Abbreviation
+
+## Table 1: Abbreviations
+
+| **Term** | **Meaning**                  |
+| -------- | ---------------------------- |
+| PMS      | Port Mac Security            |
+| SAI      | Switch Abstraction Interface |
+| FDB      | Forwarding database          |
+
+# 1 Requirements Overview
+
+## 1.1 Functional Requirements
+
+- Configuration to enable port security on a port. 
+- Configuration to set the maximum number of allowed MAC addresses on a port. 
+- Configuration to specify the action to be taken when more than the allowed number of mac addresses is attempted to be learnt on the port. 
+- Violation actions to be supported are as follows;
+	- Protect: The violating packets are dropped.
+    - Restrict: The violating packets are dropped, and a violation counter is incremented for each violation.
+    - Shutdown: The port is administratively shut down.
+
+## 1.2 Configuration and Management Requirements
+
+This feature will support REST based configurations.
+
+## 1.3 Scalability Requirements
+
+None
+
+## 1.4 Warm Boot Requirements
+
+Warm boot should be supported.
+
+# 2 Functionality
+
+## 2.1 Functional Description
+
+- Violation action to be taken when more than the allowable number MAC is learned on the port. Default action is to drop all unknown MAC traffic on the port in HW after crossing the user specified threshold (Protect).
```

Comment:
I presume this feature should behave the same for hardware-based learning or software-assisted learning? Can you please clarify?

Reply:
Yes, it should behave the same. SAI internally may change the learning mode for the port on which PMS is enabled. But, it will be transparent to SONiC.